src/libaktualizr/storage/sqlstorage.cc

```diff
@@ -1323,6 +1323,57 @@ bool SQLStorage::loadDeviceInstallationResult(data::InstallationResult* result,
   return true;
 }
 
+void SQLStorage::saveEcuReportCounter(const Uptane::EcuSerial& ecu_serial, const int64_t counter) {
+  SQLite3Guard db = dbConnection();
+
+  auto statement = db.prepareStatement<std::string, int64_t>(
+      "INSERT OR REPLACE INTO ecu_report_counter (ecu_serial, counter) VALUES "
+      "(?,?);",
+      ecu_serial.ToString(), counter);
+  if (statement.step() != SQLITE_DONE) {
+    LOG_ERROR << "Can't set ecu counter: " << db.errmsg();
+    return;
+  }
+}
+
+bool SQLStorage::loadEcuReportCounter(std::vector<std::pair<Uptane::EcuSerial, int64_t>>* results) {
+  SQLite3Guard db = dbConnection();
+
+  std::vector<std::pair<Uptane::EcuSerial, int64_t>> ecu_cnt;
+
+  // keep the same order as in ecu_serials (start with primary)
+  auto statement = db.prepareStatement(
+      "SELECT ecu_serial, counter FROM ecu_report_counter INNER JOIN ecu_serials ON "
+      "ecu_serials.serial = ecu_serial AND ecu_serials.is_primary = 1 ORDER BY ecu_serials.id;");
```

Comment:
Why the inner join? Is that just to get the Primary first?

Reply:
Yes, I want this sql help me filter the Primary for now. There is already a loop several lines below, it will put all of the ecu_serial in the vector. So, when I need to implement this feature for all of the ECUs, in this file I only need to modify this line of sql. I'm not very familiar with sql, just learned a few in 15 minutes. Do you think it's not good to use it in this way ?

Reply:
Now I get it. I still think the `"ecu_serials.serial = ecu_serial AND ecu_serials.is_primary = 1` is unnecessary, though. You might as well get everything, and ordering by `ecu_serials.id` should mean the Primary is always first. You can see a similar example in `SQLStorage::loadEcuInstallationResults()` that does about the same thing.

Reply:
It seems the SQLStorage::loadEcuInstallationResults() is also using "INNER JOIN". I think the reason might be that we only have "id" column in ecu_serial table, so if we want to order the result by "id", we still have to inner join with ecu_serial table. I think "ORDER BY" might be better than default order of "select".

Reply:
Sorry, I misquoted what I thought was unnecessary. I think the `INNER JOIN` and `ORDER BY` are fine, it's just the `AND ecu_serials.is_primary = 1` that I think is unnecessary.